src/org/mozilla/javascript/NativeArray.java

```diff
@@ -847,13 +848,17 @@ private void setLength(Object val) {
      * or its value is not convertible to a number.
      */
     static long getLengthProperty(Context cx, Scriptable obj) {
-        // These will both give numeric lengths within Uint32 range.
+        // These will give numeric lengths within Uint32 range.
         if (obj instanceof NativeString) {
             return ((NativeString)obj).getLength();
         }
         if (obj instanceof NativeArray) {
             return ((NativeArray)obj).getLength();
         }
+        if (obj instanceof XMLObject) {
+            Callable lengthFunc = (Callable) ((XMLObject) obj).get("length", obj);
```

Comment:
This seems to assume that "length" is present on all XMLObject instances and that it's a function. Is there a possibility that it is really something else, or not present at all?

Also, no big deal but the additional cast to "XMLObject" isn't doing anything here since you have to cast it to Callable anyway.

Reply:
The e4x implementation is kind of odd, and I don't really understand why it's done the way that it is. XMLObject is a public class, but does not really represent what an e4x object is. The real e4x api is in XMLObjectImpl, but that class is package-private (and in a different package,) so I can't call the length method on it directly. Instead I need to go the round-about way to get the method as a property from the object. The only two implementations of XMLObjectImpl, XML and XMLList, will both always have a length method, per the e4x spec.

I'm casting `obj` to an XMLObject, and the return value to a Callable.

Reply:
Agreed -- the XML implementation is weird but probably not worth making un-weird at this point. I think this is fine to go for now.

Reply:
Just for future reference: I think the separation of XMLObject and XMLObjectImpl stems from the time where the actual implementation of E4X relied on 3rd party libraries for all the XML handling and when not enabling E4X in Rhino the 3rd party xml libs shouldn't be a hard dependancy